gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/iap/IAPAuthenticationFilter.groovy

```diff
@@ -0,0 +1,204 @@
+import com.nimbusds.jose.JWSVerifier
+import com.nimbusds.jose.crypto.ECDSAVerifier
+import com.nimbusds.jose.jwk.ECKey
+import com.nimbusds.jose.jwk.JWK
+import com.nimbusds.jose.jwk.JWKSet
+import com.nimbusds.jwt.JWTClaimsSet
+import com.nimbusds.jwt.SignedJWT
+import groovy.util.logging.Slf4j
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+import retrofit.RetrofitError
+
+import javax.servlet.Filter
+import javax.servlet.FilterChain
+import javax.servlet.FilterConfig
+import javax.servlet.ServletException
+import javax.servlet.ServletRequest
+import javax.servlet.ServletResponse
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+import java.security.interfaces.ECPublicKey
+
+/***
+ * This filter verifies the request header from Cloud IAP containing a JWT token, after the user
+ * has been authenticated and authorized by IAP via Google OAuth2.0 and IAP's authorization service.
+ * The user email from the payload used to create the Spinnaker user.
+ */
+@Slf4j
+class IAPAuthenticationFilter implements Filter {
+
+  IAPConfig.IAPSecurityConfigProperties configProperties
+
+  PermissionService permissionService
+
+  Front50Service front50Service
+
+  private static final String signatureAttribute = "JWTSignature"
+
+  private final Map<String, JWK> keyCache = new HashMap<>()
+  private final Logger logger = LoggerFactory.getLogger(IAPAuthenticationFilter.class)
+
+  IAPAuthenticationFilter(
+    IAPConfig.IAPSecurityConfigProperties configProperties,
+    PermissionService permissionService,
+    Front50Service front50Service) {
+    this.configProperties = configProperties
+    this.permissionService = permissionService
+    this.front50Service = front50Service
+  }
+
+  @Override
+  void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    HttpServletRequest req = (HttpServletRequest) request
+    HttpServletResponse res = (HttpServletResponse) response
+
+
+    try {
+      def token = req.getHeader(configProperties.jwtHeader)
+      Preconditions.checkNotNull(token)
+
+      SignedJWT jwt = SignedJWT.parse(token)
+
+
+      def signatureInSession = req.getSession(false)?.getAttribute(signatureAttribute)
+
+
+      if (signatureInSession && signatureInSession == jwt.signature) {
+        // Signature matches in previous request signatures in current session, skip validation.
+        chain.doFilter(req, res)
+        return
+      }
+
+      def verifiedUser = verifyJWTAndGetUser(jwt)
+
+      PreAuthenticatedAuthenticationToken authentication = new PreAuthenticatedAuthenticationToken(
+        verifiedUser,
+        null /* credentials */,
+        verifiedUser.authorities
```

Comment:
I'd add a comment that you have to provide the `authorities` list (even though it's empty) in order to get an object with `authenticated = true`

Reply:
Done.